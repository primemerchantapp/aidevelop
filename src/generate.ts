import type { Project, EmbedOptions, OpenOptions } from './interfaces';
import { PROJECT_TEMPLATES } from './constants';
import { embedUrl, openTarget, openUrl } from './helpers';

function createHiddenInput(name: string, value: string) {
  const input = document.createElement('input');
  input.type = 'hidden';
  input.name = name;
  input.value = value;
  return input;
}

/**
 * Encode file paths for use in input name attributes.
 * We need to replace square brackets (as used by Next.js, SvelteKit, etc.),
 * with custom escape sequences. Important: do not encodeURIComponent the
 * whole path, for compatibility with the StackBlitz backend.
 */
function encodeFilePath(path: string) {
  return path.replace(/\[/g, '%5B').replace(/\]/g, '%5D');
}

export function createProjectForm({
  template,
  title,
  description,
  dependencies,
  files,
  settings,
}: Project) {
  if (!PROJECT_TEMPLATES.includes(template)) {
    const names = PROJECT_TEMPLATES.map((t) => `'${t}'`).join(', ');
    console.warn(`Unsupported project.template: must be one of ${names}`);
  }

  const inputs: HTMLInputElement[] = [];
  const addInput = (name: string, value: string, defaultValue = '') => {
    inputs.push(createHiddenInput(name, typeof value === 'string' ? value : defaultValue));
  };

  addInput('project[title]', title);
  if (typeof description === 'string' && description.length > 0) {
    addInput('project[description]', description);
  }
  addInput('project[template]', template, 'javascript');

  if (dependencies) {
    if (template === 'node') {
      console.warn(
        `Invalid project.dependencies: dependencies must be provided as a 'package.json' file when using the 'node' template.`
      );
    } else {
      addInput('project[dependencies]', JSON.stringify(dependencies));
    }
  }

  if (settings) {
    addInput('project[settings]', JSON.stringify(settings));
  }

  Object.entries(files).forEach(([path, contents]) => {
    addInput(`project[files][${encodeFilePath(path)}]`, contents);
  });

  const form = document.createElement('form');
  form.method = 'POST';
  form.setAttribute('style', 'display:none!important;');
  form.append(...inputs);
  return form;
}

export function createProjectFrameHTML(project: Project, options?: EmbedOptions) {
  const form = createProjectForm(project);
  form.action = embedUrl('/run', options);
  form.id = 'sb_run';

  const html = `<!doctype html>
<html>
<head><title></title></head>
<body>
  ${form.outerHTML}
  <script>document.getElementById('${form.id}').submit();</script>
</body>
</html>`;

  return html;
}

export function openNewProject(project: Project, options?: OpenOptions) {
  const form = createProjectForm(project);
  form.action = openUrl('/run', options);
  form.target = openTarget(options);

  document.body.appendChild(form);
  form.submit();
  document.body.removeChild(form);
}
