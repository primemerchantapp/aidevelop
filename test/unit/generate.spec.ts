/** @vitest-environment happy-dom */
import { afterEach, describe, expect, test, vi } from 'vitest';

import { createProjectForm, createProjectFrameHTML, openNewProject } from '$src/generate';
import { useMockConsole, restoreConsole } from './utils/console';
import { formValue, makeContainer, removeContainer } from './utils/dom';
import { getTestProject } from './utils/project';

const cleanUp = () => {
  restoreConsole();
  removeContainer();
};

describe('createProjectFrameHTML', () => {
  afterEach(cleanUp);

  test('generates a HTML document string', () => {
    const html = createProjectFrameHTML(getTestProject());
    expect(html).toBeTypeOf('string');
    expect(html).toMatchSnapshot();
  });

  test('generated HTML contains a form', () => {
    const parent = makeContainer();
    parent.innerHTML = createProjectFrameHTML(getTestProject(), {
      view: 'editor',
      terminalHeight: 50,
    });

    // Check for form element
    const form = parent.querySelector('form')!;
    expect(form).toBeInstanceOf(HTMLFormElement);
    expect(form.getAttribute('action')).toBe(
      'https://stackblitz.com/run?embed=1&view=editor&terminalHeight=50'
    );
  });
});

describe('createProjectForm', () => {
  afterEach(cleanUp);

  test('generates a form element', () => {
    const project = getTestProject({
      title: 'My Test Project',
      dependencies: { cowsay: '1.5.0' },
      settings: { compile: { clearConsole: false } },
    });
    const form = createProjectForm(project);
    const value = (name: string) => formValue(form, name);

    // Check input values
    expect(value('project[title]')).toBe('My Test Project');
    expect(value('project[description]')).toBe(undefined);
    expect(value('project[template]')).toBe(project.template);
    expect(value('project[dependencies]')).toBe(JSON.stringify(project.dependencies));
    expect(value('project[settings]')).toBe(JSON.stringify(project.settings));
    expect(value('project[files][package.json]')).toBe(project.files['package.json']);
    expect(value('project[files][index.js]')).toBe(project.files['index.js']);
  });

  test('warns on unknown template type', () => {
    const console = useMockConsole();
    const warnSpy = vi.spyOn(console, 'warn');

    const project = getTestProject({ template: 'doesntexist' as any });
    const form = createProjectForm(project);
    const value = (name: string) => formValue(form, name);

    // Unknown template value is accepted by triggers a warning
    expect(value('project[template]')).toBe('doesntexist');
    expect(warnSpy).toHaveBeenCalledWith(
      `Unsupported project.template: must be one of 'angular-cli', 'create-react-app', 'html', 'javascript', 'node', 'polymer', 'typescript', 'vue'`
    );
  });

  test('ignores dependencies for node template', () => {
    const console = useMockConsole();
    const warnSpy = vi.spyOn(console, 'warn');

    const project = getTestProject({
      template: 'node',
      dependencies: { serve: '^12' },
    });
    const form = createProjectForm(project);
    const value = (name: string) => formValue(form, name);

    // Unknown template value is accepted by triggers a warning
    expect(value('project[template]')).toBe('node');
    expect(value('project[dependencies]')).toBe(undefined);
    expect(warnSpy).toHaveBeenCalledWith(
      `Invalid project.dependencies: dependencies must be provided as a 'package.json' file when using the 'node' template.`
    );
  });
});

describe('openNewProject', () => {
  test('cleans up after itself', () => {
    const project = getTestProject();

    expect(document.body.children.length).toBe(0);
    expect(() => openNewProject(project)).not.toThrow();
    expect(document.body.children.length).toBe(0);
  });
});
