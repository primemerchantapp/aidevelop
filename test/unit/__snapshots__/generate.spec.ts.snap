// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createProjectFrameHTML > generates a HTML document string 1`] = `
"<!doctype html>
<html>
<head><title></title></head>
<body>
  <form method=\\"POST\\" style=\\"display:none!important;\\" action=\\"https://stackblitz.com/run?embed=1\\" id=\\"sb_run\\"><input type=\\"hidden\\" name=\\"project[title]\\" value=\\"Test Project\\"><input type=\\"hidden\\" name=\\"project[template]\\" value=\\"javascript\\"><input type=\\"hidden\\" name=\\"project[files][package.json]\\" value=\\"{
  &quot;dependencies&quot;: {&quot;cowsay&quot;: &quot;1.5.0&quot;}
}\\"><input type=\\"hidden\\" name=\\"project[files][index.js]\\" value=\\"import cowsay from &#x27;cowsay&#x27;;
console.log(cowsay(&#x27;Hello world!&#x27;));
\\"></form>
  <script>document.getElementById('sb_run').submit();</script>
</body>
</html>"
`;
